File: web/vroom/src/app/core/services/ride.service.ts
import { HttpClient } from "@angular/common/http";
import { Injectable } from "@angular/core";
import { CancelRideRequestDTO } from "../models/ride/requests/cancel-ride-req.dto";
import { MessageResponseDTO } from "../models/message-response.dto";
import { Observable } from "rxjs";
import { StopRideRequestDTO } from "../models/ride/requests/stop-ride-req.dto";
import { StoppedRideResponseDTO } from "../models/ride/responses/stopped-ride-response.dto";
import { LeaveReviewRequestDTO } from "../models/ride/requests/leave-review-req.dto";
import { ComplaintRequestDTO } from "../models/ride/requests/complaint-req.dto";
import { GetRouteResponseDTO } from "../models/ride/responses/get-route-response.dto";
import { map } from "rxjs/operators";
import { HttpHeaders } from "@angular/common/http";
import { GetRideResponseDTO } from '../models/ride/responses/get-ride-response.dto'

@Injectable({
    providedIn: "root"
})
export class RideService{
    private rideUrl = 'http://localhost:8080/api/rides'

    constructor(private http: HttpClient) {}

    getUserRide(): Observable<GetRideResponseDTO> {
        return this.http.get<GetRideResponseDTO>(this.rideUrl+"/user-active-ride")
    }

    cancelRideRequest(rideID: string, data: CancelRideRequestDTO): Observable<MessageResponseDTO>{
        return this.http.put<MessageResponseDTO>(this.rideUrl+`/${rideID}`+'/cancel', data);
    }

    stopRideRequest(rideID: string, data: StopRideRequestDTO): Observable<StoppedRideResponseDTO>{
        return this.http.put<StoppedRideResponseDTO>(this.rideUrl+`/${rideID}`+'/stop', data);
    }

    leaveReviewRequest(rideID: string, data: LeaveReviewRequestDTO): Observable<MessageResponseDTO>{
        console.log(data);
        return this.http.post<MessageResponseDTO>(this.rideUrl+`/${rideID}`+'/review', data)
    }

    finishRideRequest(rideID: string): Observable<MessageResponseDTO>{
        return this.http.post<MessageResponseDTO>(this.rideUrl+`/${rideID}`+'/finish', null)
    }

    sendComplaintRequest(rideID: string, data: ComplaintRequestDTO): Observable<MessageResponseDTO>{
        return this.http.post<MessageResponseDTO>(this.rideUrl+'/complaint'+`/${rideID}`, data)
    }

    getRouteDetails(rideID: string): Observable<GetRouteResponseDTO>{
        console.log('Fetching route details for rideID:', this.rideUrl+`/${rideID}`+'/route');
        return this.http.get<GetRouteResponseDTO>(this.rideUrl+'/route'+`/${rideID}`)
    }

    getAddress(lat: number, lng: number): Observable<string> {
        const url = `https://nominatim.openstreetmap.org/reverse?format=json&lat=${lat}&lon=${lng}`;
        const headers = new HttpHeaders().set('User-Agent', 'Vroom (korda.isidora@gmail.com)');
        return this.http.get<any>(url, { headers }).pipe(
        map(res => {
                if (!res || !res.display_name) return 'Unknown Address';
                const parts = res.display_name.split(',');
                return parts.slice(0, 2).map((p: string) => p.trim()).join(', ');
            })
        );
    }

    
}

File: mobile/app/src/main/java/com/example/vroom/services/RideService.java
package com.example.vroom.services;

import com.example.vroom.DTOs.MessageResponseDTO;
import com.example.vroom.DTOs.ride.requests.CancelRideRequestDTO;
import com.example.vroom.DTOs.ride.requests.FavoriteRouteDTO;
import com.example.vroom.DTOs.ride.requests.LeaveReviewRequestDTO;
import com.example.vroom.DTOs.ride.requests.OrderFromFavoriteRequestDTO;
import com.example.vroom.DTOs.ride.requests.RideRequestDTO;
import com.example.vroom.DTOs.ride.requests.ComplaintRequestDTO;
import com.example.vroom.DTOs.ride.requests.StopRideRequestDTO;
import com.example.vroom.DTOs.ride.responses.GetRideResponseDTO;
import com.example.vroom.DTOs.ride.responses.RideHistoryResponseDTO;
import com.example.vroom.DTOs.ride.responses.StoppedRideResponseDTO;
import com.example.vroom.DTOs.ride.responses.UserActiveRideDTO;
import com.example.vroom.DTOs.route.responses.GetRouteResponseDTO;

import java.util.List;

import retrofit2.Call;
import retrofit2.http.Body;
import retrofit2.http.DELETE;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.PUT;
import retrofit2.http.Path;
import retrofit2.http.Query;

public interface RideService {

    @PUT("api/rides/{rideID}/cancel")
    Call<MessageResponseDTO> cancelRide(
            @Path("rideID") Long rideId,
            @Body CancelRideRequestDTO data
    );

    @PUT("api/rides/{rideID}/stop")
    Call<StoppedRideResponseDTO> stopRide(
            @Path("rideID") Long rideId,
            @Body StopRideRequestDTO data
    );

    @GET("api/rides/active")
    Call<List<GetRideResponseDTO>> getActiveRides();

    @PUT("api/rides/start/{rideID}")
    Call<GetRideResponseDTO> startRide(@Path("rideID") Long rideID);

    @POST("api/rides")
    Call<GetRideResponseDTO> orderRide(@Body RideRequestDTO request);

    @GET("api/rides/favorites")
    Call<List<FavoriteRouteDTO>> getFavoriteRoutes();

    @POST("api/rides/order/favorite")
    Call<GetRideResponseDTO> orderFromFavorite(@Body OrderFromFavoriteRequestDTO request);

    @DELETE("api/rides/favorites/{favoriteId}")
    Call<MessageResponseDTO> removeFavorite(@Path("favoriteId") Long favoriteId);
    @GET("api/rides/user-active-ride")
    Call<List<UserActiveRideDTO>> getUserActiveRide();

    @GET("api/rides/route/{rideID}")
    Call<GetRouteResponseDTO> getRoute(
            @Path("rideID") Long rideID
    );

    @POST("api/rides/complaint/{rideID}")
    Call<MessageResponseDTO> sendComplaint(
            @Path("rideID") Long rideID,
            @Body ComplaintRequestDTO complaint
    );

    @POST("api/rides/{rideID}/finish")
    Call<MessageResponseDTO> finishRide(
            @Path("rideID") Long rideID
    );

    @POST("api/rides/{rideID}/review")
    Call<MessageResponseDTO> leaveReview(
            @Path("rideID") Long rideID,
            @Body LeaveReviewRequestDTO body
    );
}
